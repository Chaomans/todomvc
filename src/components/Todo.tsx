import React, { useState } from "react";

type todoProps = {
    description: string;
    onDelete: () => void;
    onChangeDone: () => void;
    onEditDescription: (desc: string) => void;
    isdone: boolean;
};

const Todo = ({
    description,
    onEditDescription,
    onDelete,
    onChangeDone,
    isdone,
}: todoProps) => {
    const [done, setDone] = useState<boolean>(isdone);
    const [editing, setEditing] = useState<boolean>(false);
    const [newDesc, setNewDesc] = useState<string>(description);

    const onClickDone = () => {
        setDone(!done);
        onChangeDone();
    };

    const onDoubleClick = () => {
        setEditing(true);
    };

    const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setNewDesc(e.target.value);
    };

    const replaceDesc = (e: React.KeyboardEvent<HTMLInputElement>) => {
        if (e.key === "Enter" && newDesc.length > 3) {
            onEditDescription(newDesc);
            setEditing(false);
        }
    };

    const setLiClass = (): string => {
        const liClass = [];
        if (done) liClass.push("completed");
        if (editing) liClass.push("editing");
        return liClass.join(" ");
    };

    return (
        <li className={setLiClass()}>
            <div className="view">
                <input
                    type="checkbox"
                    className="toggle"
                    onChange={onClickDone}
                    checked={done}
                ></input>
                <label onDoubleClick={onDoubleClick}>{description}</label>
                <button onClick={() => onDelete()} className="destroy"></button>
            </div>
            <input
                type="text"
                className="edit"
                autoFocus
                onKeyDown={replaceDesc}
                onChange={onChange}
                value={newDesc}
            />
        </li>
    );
};

export default Todo;
