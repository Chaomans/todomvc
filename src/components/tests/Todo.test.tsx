import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Todo from "../Todo";

describe("Todo component", () => {
    it("should render properly", () => {
        const { asFragment } = render(
            <Todo
                description="First thing to do"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={() => {}}
                onEditDescription={() => {}}
                onSetEditing={() => {}}
            />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it.each`
    completed   | status    
    ${true}     | ${"be"}
    ${false}     | ${"not be"}
    `("Should $status checked if completed", async ({completed}) => {
        render(
            <Todo
                description={"todo"}
                isdone={completed}
                onChangeDone={() => {}}
                onDelete={() => {}}
                onEditDescription={() => {}}
                onSetEditing={() => {}}
            />
        );

        const checkbox = await screen.findByRole("checkbox") as HTMLInputElement;
        expect(checkbox.checked).toEqual(completed);
    });

    it.each`
    completed   | status
    ${true}     | ${"completed"}
    ${false}    | ${"active"}
    `("Should notify the parent when change status from $status", async ({completed}) => {
        const mockOnChangeDone = jest.fn();
        render(
            <Todo
                description="First thing to do"
                isdone={completed}
                onChangeDone={mockOnChangeDone}
                onDelete={() => {}}
                onEditDescription={() => {}}
                onSetEditing={() => {}}
            />
        );

        await userEvent.click(await screen.findByRole("checkbox"));

        expect(mockOnChangeDone).toHaveBeenCalledTimes(1);
        expect(mockOnChangeDone).toHaveBeenCalledWith(!completed);
    });

    it("Should notify parent on enter in editing mode", async () => {
        const mockOnSetEditing = jest.fn();
        render(
            <Todo
                description="to edit"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={() => {}}
                onEditDescription={() => {}}
                onSetEditing={mockOnSetEditing}
            />
        );

        await userEvent.dblClick(await screen.findByText("to edit"));

        expect(mockOnSetEditing).toHaveBeenCalledTimes(1);
        expect(mockOnSetEditing).toHaveBeenCalledWith(true);
    });

    it("Should switch between view and edit on enter and on exit editing mode", async () => {
        const mockOnSetEditing = jest.fn();
        render(
            <Todo
                description="to do"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={() => {}}
                onEditDescription={() => {}}
                onSetEditing={mockOnSetEditing}
            />
        );

        // enter: hide view, show edit
        await userEvent.dblClick(await screen.findByText("to do"));
        expect(screen.queryByText("to do")).not.toBeInTheDocument();
        expect(await screen.findByRole("textbox")).toBeInTheDocument();
        // exit: show view, hide edit
        await userEvent.keyboard('{Escape}');
        expect(await screen.findByText("to do")).toBeInTheDocument();
        expect(screen.queryByRole("textbox")).not.toBeInTheDocument();
    })

    it("Should notify parent if edit is valid", async () => {
        const mockOnEditDescription = jest.fn();
        const mockOnSetEditing = jest.fn();
        render(
            <Todo
                description="to edit"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={() => {}}
                onEditDescription={mockOnEditDescription}
                onSetEditing={mockOnSetEditing}
            />
        );
        await userEvent.dblClick(await screen.findByText("to edit"));
        expect(mockOnSetEditing).toHaveBeenNthCalledWith(1, true);
        await userEvent.type( await screen.findByRole("textbox"), "insert{Enter}");
        expect(mockOnEditDescription).toHaveBeenCalledTimes(1);
        expect(mockOnEditDescription).toHaveBeenCalledWith("to editinsert");
        expect(mockOnSetEditing).toHaveBeenNthCalledWith(2, false);
    });

    it("Should notify parent to delete if edit is invalid", async () => {
        const mockOnDelete = jest.fn();
        const mockOnSetEditing = jest.fn();
        render(
            <Todo
                description="to edit"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={mockOnDelete}
                onEditDescription={() => {}}
                onSetEditing={mockOnSetEditing}
            />
        );
        await userEvent.dblClick(await screen.findByText("to edit"));
        expect(mockOnSetEditing).toHaveBeenNthCalledWith(1, true);
        await userEvent.type( await screen.findByRole("textbox"),`${'{backspace}'.repeat(7)} \t{Enter}`);
        expect(mockOnDelete).toHaveBeenCalledTimes(1);
        expect(mockOnDelete).toHaveBeenCalledWith();
    });

    it("Should notify parent when trying to delete", async () => {
        const mockDelete = jest.fn();
        render(
            <Todo
                description="First thing to do"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={mockDelete}
                onEditDescription={() => {}}
                onSetEditing={() => {}}
            />
        );

        await userEvent.click(await screen.findByRole("button"));

        expect(mockDelete).toHaveBeenCalledTimes(1);
        expect(mockDelete).toHaveBeenCalledWith();

    });

    it("Should discard changes and exit editing mode when escape is pressed", async () => {
        const mockOnEditDescription = jest.fn();
        const mockOnSetEditing = jest.fn();
        render(
            <Todo
                description="First thing to do"
                isdone={false}
                onChangeDone={() => {}}
                onDelete={() => {}}
                onEditDescription={mockOnEditDescription}
                onSetEditing={mockOnSetEditing}
            />
        );

        await userEvent.dblClick(await screen.findByText("First thing to do"));
        expect(mockOnSetEditing).toHaveBeenCalledTimes(1);
        expect(mockOnSetEditing).toHaveBeenCalledWith(true);
        await userEvent.type(await screen.findByRole("textbox"), "insert");
        await userEvent.keyboard('{Escape}');
        expect(mockOnEditDescription).not.toHaveBeenCalled();
        expect(mockOnSetEditing).toHaveBeenLastCalledWith(false);

    });   
    
});
