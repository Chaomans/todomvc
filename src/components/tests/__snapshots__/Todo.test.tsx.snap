// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo component should render properly 1`] = `
<DocumentFragment>
  <div
    class="view"
  >
    <input
      class="toggle"
      type="checkbox"
    />
    <label>
      First thing to do
    </label>
    <button
      class="destroy"
    />
  </div>
</DocumentFragment>
`;
